### PR TITLE
Mk 235887 configure menu grid

### PR DESCRIPTION
**Pull request by mkangia**
http://manage.dimagi.com/default.asp?235887

### CODE REVIEW

**Comment by czue** [comment]
fyi, lots of different build failures. you probably want to run all the app manager tests locally until they pass

```
======================================================================
ERROR: test_basic_workflow (corehq.apps.app_manager.tests.test_child_module.BasicModuleAsChildTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_child_module.py", line 49, in test_basic_workflow
    self.assertXmlPartialEqual(XML, self.app.create_suite(), "./menu")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5215, in create_suite
    return SuiteGenerator(self, build_profile_id).generate_suite()
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/generator.py", line 60, in generate_suite
    menus.get_module_contributions(module)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/sections/menus.py", line 130, in get_module_contributions
    if self.app.grid_display_for_all_modules():
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5652, in grid_display_for_all_modules
    return self.profile['properties']['cc-grid-menus'] == 'all'
  File "/vendor/lib/python2.7/site-packages/jsonobject/containers.py", line 148, in __getitem__
    return super(JsonDict, self).__getitem__(key)
KeyError: 'properties'
```
```
======================================================================
ERROR: test_advanced_suite (corehq.apps.app_manager.tests.test_advanced_suite.AdvancedSuiteTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_advanced_suite.py", line 26, in test_advanced_suite
    self._test_generic_suite('suite-advanced')
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 72, in _test_generic_suite
    self.assertXmlEqual(self.get_xml(suite_tag), app.create_suite())
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5215, in create_suite
    return SuiteGenerator(self, build_profile_id).generate_suite()
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/generator.py", line 60, in generate_suite
    menus.get_module_contributions(module)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/sections/menus.py", line 130, in get_module_contributions
    if self.app.grid_display_for_all_modules():
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5652, in grid_display_for_all_modules
    return self.profile['properties']['cc-grid-menus'] == 'all'
  File "/vendor/lib/python2.7/site-packages/jsonobject/containers.py", line 148, in __getitem__
    return super(JsonDict, self).__getitem__(key)
KeyError: 'cc-grid-menus'
```
```
======================================================================
ERROR: test_case_list_lookup_autolaunch (corehq.apps.app_manager.tests.test_case_list_lookup.CaseListLookupTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_case_list_lookup.py", line 72, in test_case_list_lookup_autolaunch
    app.create_suite(),
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5215, in create_suite
    return SuiteGenerator(self, build_profile_id).generate_suite()
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/generator.py", line 60, in generate_suite
    menus.get_module_contributions(module)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/suite_xml/sections/menus.py", line 130, in get_module_contributions
    if self.app.grid_display_for_all_modules():
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5652, in grid_display_for_all_modules
    return self.profile['properties']['cc-grid-menus'] == 'all'
  File "/vendor/lib/python2.7/site-packages/jsonobject/containers.py", line 148, in __getitem__
    return super(JsonDict, self).__getitem__(key)
KeyError: 'properties'
```

**Comment by mkangia** [comment]
@snopoke @orangejenny @wspride : Any review points?

File: corehq/apps/app_manager/templates/app_manager/module_view.html [hunk 1]
```diff
@@ -305,6 +305,22 @@
...
                 </div>
+
+                {% if app.grid_display_for_some_modules %}
+                <div class="form-group">
+                    <label class="col-sm-2 control-label">
+                        {% trans "Display Style" %}
+                    </label>
+                    <div class="col-sm-4">
+                        <select type="text" name="display_style" class="form-control">
+                            <option value="list">{% trans "list(the default)" %}</option>
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/module_view.html]: `list(the default)` should just be `List`

File: corehq/apps/app_manager/templates/app_manager/module_view.html [hunk 2]
```diff
@@ -305,6 +305,22 @@
                         </select>
                     </div>
                 </div>
+
+                {% if app.grid_display_for_some_modules %}
+                <div class="form-group">
+                    <label class="col-sm-2 control-label">
+                        {% trans "Display Style" %}
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/module_view.html]: Is there any description of grid view on confluence? (Jeremy or Amelia would know.) Would be nice to have a help bubble that links to it - this is one of those features that's obviously great if you can see a sample image, but hard to describe in words what it does.

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml [hunk 3]
```diff
@@ -350,10 +350,10 @@
 - name: "Grid Menus"
   description: "If enabled module and form menus will display in a grid view instead of a list."
   id: "cc-grid-menus"
-  values: ["yes", "no"]
-  value_names: ["Yes", "No"]
-  default: "no"
-  disabled_default: "no"
+  values: ["none", "first", "some", "all"]
+  value_names: ["None", "First Screen", "Some Screens", "All Screens"]
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml]: "Some screens" is rather ambiguous. It looks like there was discussion in the spec about having a hint as to what that means. Was that resolved?

Could add something to the description, or change "Some Screens" to something like "Customize per Module."

**Comment by orangejenny** [comment]
Looks reasonable. Making some comments, but they're all minor.

Would be nice to add tests to https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/tests/test_grid_menus.py

File: corehq/apps/app_manager/templates/app_manager/module_view.html (hunk 2 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/module_view.html]: Will check with Amelia how the user can be best guided on how to use this. Sample images might be good for Customisation for each module.

**Comment by czue** [comment]
@mkangia did you see that @amsagoff and @wpride updated the spec a little bit?

**Comment by mkangia** [comment]
@czue : Yup! we have been following up discussion on that via email. Looping you in.

File: corehq/apps/app_manager/static/app_manager/js/commcaresettings.js [hunk 4]
```diff
@@ -236,6 +236,7 @@ hqDefine('app_manager/js/commcaresettings.js', function () {
             setting.hasError = ko.computed(function () {
                 return setting.disabledButHasValue() || !setting.valueIsLegal();
             });
+            setting.foot_note = ko.observable(setting.foot_note)
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/commcaresettings.js]: Missing semicolon. (semi)

File: corehq/apps/app_manager/models.py [hunk 5]
```diff
@@ -5639,6 +5642,17 @@ def get_subcase_types(self, case_type):
                 if m.case_type == case_type
                 for t in m.get_subcase_types()}
 
+    def grid_display_for_some_modules(self):
+        return self.cc_grid_menus_property == 'some'
+
+
+    def grid_display_for_all_modules(self):
```
Comment (dimagimon) [inline corehq/apps/app_manager/models.py]: E303 too many blank lines (2)

**Comment by mkangia** [comment]
@amsagoff : Screenshots to confirm the text changes.
screen shot 2016-09-10 at 11 24 49 pm

screen shot 2016-09-10 at 11 25 28 pm

screen shot 2016-09-10 at 11 25 36 pm

Also, jenny suggested to provide link to a help doc on confluence. Can add that though now it seems quite straight forward and easy to understand.

**Comment by amsagoff** [comment]
looks good, thanks!

**Comment by mkangia** [comment]
@orangejenny feedback on changes please.

**Comment by orangejenny** [comment]
Thanks for adding tests. Code changes look good.

I do still think it'd be good to have a public confluence page that explains what grid menus are, with screenshots, and then link to that page from the help bubbles. I'll ping documentation@dimagi to see if they think it's worth making such a page. That doesn't need to block this.

**Pull request merged**